File: leetcode/2-add-numbers.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode nHead = new ListNode();
        ListNode curr = nHead;
        int carry = 0;

        while (l1 != null || l2 != null || carry != 0) {
            int num1 = (l1 != null) ? l1.val : 0;
            int num2 = (l2 != null) ? l2.val : 0;

            int sum = num1 + num2 + carry;
            int actual = sum % 10;
            carry = sum / 10; // integer truncated downwards towards zero

            curr.next = new ListNode(actual);
            curr = curr.next;

            l1 = (l1 != null) ? l1.next : null;
            l2 = (l2 != null) ? l2.next : null;

            // if end and still carry, 0 + 0 + 1, next is 0 0 0 which loop doesnt start
        }
        return nHead.next;
        

    }
}

File: leetcode/2-add-numbers.js
const addTwoNumbers = (l1, l2) => {
    let nHead = new ListNode();
    let curr = nHead;
    let carry = 0;

    while (l1 || l2 || carry) {
        // conditional nums -> && syntax is var = this () if this ()
        let n1 = l1 && l1.val;
        let n2 = l2 && l2.val;

        // get total
        let sum = n1 + n2 + carry;

        // set next carry
        carry = Math.floor(sum/10);

        // create and set next node floor div
        curr.next = new ListNode(sum%10);

        // move pointers -> else leave at null
        if (l1) l1 = l1.next;
        if (l2) l2 = l2.next;

        curr = curr.next;

    }

    return nHead.next;
}